Add kotlin guide post (#113)
oktadeveloper/okta-blog: _source/_posts/2019-12-12-tutorial-kotlin-beginners-guide.md

---
layout: blog_post
title: "Kotlin: A Beginner's Guide and Tutorial"
author: jimena-garbarino

> description: "Tutorial: Learn the basics of Kotlin by building an app, find out why Java developers love Kotlin!"

Reviewer: ```suggestion
description: "Tutorial: Learn the basics of Kotlin by building an app and find out why Java developers love Kotlin!"
```

tags: [kotlin, java, tutorial, oauth2, oidc, guide]
tweets:
- "Find out why Java developers ❤️ @Kotlin in this tutorial!"
- "Learn the basics of @Kotlin and build a secure app with @SpringBoot 🛠️"
- "Haven't tried @Kotlin yet? Check out this tutorial!"
image: blog/tutorial-kotlin-beginners-guide/kotlin-logo.png
---

Kotlin is a modern, statically typed language within the JVM. Kotlin is a cross-platform, multi-purpose, free and open-source language developed by JetBrains under the Apache 2.0 license and has constructs for both Object Oriented and Functional programming styles, which can be mixed. It can be used for web development, server and client, and mobile development, using most Java IDEs.

Kotlin is an awesome option for Java developers because it is concise, expressive, and safe. According to JetBrains estimates, it can cut the total lines of code in your app by up to 40%. Kotlin also helps prevent `NullPointerException`s as it provides **non-nullable** types

> According to [Github Octoverse Report 2018](https://octoverse.github.com/projects#languages), Kotlin is the number one fastest-growing language. With 2.2M users, its popularity increases every month and big companies are using it. Android and JVM developers are falling in love with its features. Let's learn why.

Reviewer: Github > GitHub

In this post you will find:
- An overview of Kotlin's features
- A guide for building a "Hello, World!" app and solving a basic algorithm problem
- A quick step by step for running a Kotlin Spring Boot application with authentication via Okta

In order to complete the full tutorial, you'll need the following tools:
- [Java 8](https://adoptopenjdk.net/)
- [Apache Maven](https://maven.apache.org/)

> - [IntelliJ](https://www.jetbrains.com/idea/) (or other IDE)

Reviewer: ```suggestion
- [IntelliJ](https://www.jetbrains.com/idea/) (or another IDE)
```

**Use-site variance**: Java's type system uses bounded wildcards to increase API flexibility, as generic types are invariant (as opposed to Java arrays, which are covariant), meaning `List<String>` is not a subtype of `List<Object>`, but can be assigned to `List<? extends Object>` type. In Kotlin, instead of bounded wildcards, use-site variance allows to restrict the generic type in the place it is used, with a simpler syntax like `Array<out Any>`, equivalent to Java's `Array<? extends Object>`. The compiler verifies the parameter type is only returned from a method of the instance, but not consumed, to avoid runtime exceptions, for example, if attempting to write a `String` to `Array<Int>`. 

```kotlin
package com.okta.developer

fun main() {

> 

Reviewer: Remove empty blank line

    val a = arrayOf("uno", "dos", "tres")
    readMethod(a)
}

fun readMethod(a: Array<out Any>) {
    println(a[0])
    a.set(0, 1) // Out-projected type prohibits the use of fun set
}
```

> The example above does not compile in Kotlin, because the `readMethod` receives the out-projected array (only read operations allowed) and it is calling the write operation `set`.

Reviewer: ```suggestion
The example above does not compile in Kotlin, because the `readMethod()` receives the out-projected array (only read operations allowed) and it is calling the write operation `set()`.
```

**No checked exceptions**: Java checked exceptions must be somehow handled for the program to compile, and are many times swallowed by an empty catch block. Kotlin does not have checked exceptions, because it is suggested that in large software projects it decreases productivity.

**Proper function types**: As opposed to Java SAM (Single Abstract Method) conversions, where a lambda expression is convertible to a SAM type according to some rules, Kotlin uses a family of function types like `(Int) -> String`, with the special notation corresponding to the signature of the functions (parameters and return values).

A complete list of Kotlin features, not supported in Java, is available at [Kotlin Reference](https://kotlinlang.org/docs/reference/comparison-to-java.html#what-kotlin-has-that-java-does-not).

## Java and Kotlin Interoperability

Kotlin is 100% interoperable with Java, Kotlin code can be called from Java and vice versa.
Existing Java code can be used with some considerations, for example, Java getters and setters are represented by properties in Kotlin:

```kotlin
val calendar = Calendar.getInstance()
calendar.firstDayOfWeek = Calendar.MONDAY // setFirstOfWeek()
```

> With an IDE like IntelliJ IDEA, you can add Java source code to a Kotlin project just by creating the `.java` file.

Reviewer: ```suggestion
With an IDE like IntelliJ IDEA, you can add Java source code to a Kotlin project just by creating a `.java` file.
```

> Objects coming from Java, called **platform types**, have relaxed null-checks for practical reasons, and safety is the same as in Java. Kotlin will not inform a compilation error but the call might fail at runtime. For example, if Kotlin calls Java code that returns an ArrayList, the inferred type in Kotlin will be ArrayList<String!>!, which means the collection can be nullable, and the items as well.

Reviewer: ```suggestion
Objects coming from Java, called **platform types**, have relaxed null-checks for practical reasons, and safety is the same as in Java. Kotlin will not inform a compilation error but the call might fail at runtime. For example, if Kotlin calls Java code that returns an `ArrayList`, the inferred type in Kotlin will be `ArrayList<String!>!`, which means the collection can be nullable, and the items as well.
```

In the Java code below, `null` is added as an item to the list.

```java
import java.util.ArrayList;

public class ArrayWithNulls {

    public static ArrayList<String> create(){
        ArrayList<String> strings = new ArrayList();
        strings.add(null);
        strings.add("foo");
        return strings;
    }
}
```

> In the Kotlin lines below, we are calling the previous Java code, so the safety is relaxed. The program compiles, but as the first item is null, it will fail at runtime with IllegalStateException. 

Reviewer: ```suggestion
In the Kotlin lines below, we are calling the previous Java code, so the safety is relaxed. The program compiles, but as the first item is `null`, it will fail at runtime with `IllegalStateException`. 
```

```kotlin
    val list = ArrayWithNulls.create()
    val item = list.get(0) // platform type inferred (ordinary Java object)

>     item.substring(1) // IllegalStateException: item must not be null

Reviewer: Outdent this whole code block.

```

## Pre-Tutorial: A Taste of Kotlin

To avoid the programmer's curse, let's not skip trying a simple "Hello, World!" example with Kotlin. An easy way to get started with Kotlin is to download and install [IntelliJ IDEA Community Edition](https://www.jetbrains.com/idea/download/) from JetBrains, which is also free. And also install [Maven](https://maven.apache.org/download.cgi) for dependency management and to run the application from the command line.

Start IntelliJ IDEA and create a new Kotlin project, choosing **Kotlin/JVM**.

{% img blog/tutorial-kotlin-beginners-guide/intellij-new-project.png alt:"Create new IntelliJ project" width:"600" %}{: .center-image }

> Set the project name, location, and Java 8 SDK. After the project is created, add Maven support by doing a right-click on the project and choosing *Add Framework Support*. Then select the Maven box and *Ok*.

Reviewer: ```suggestion
Set the project name, location, and Java 8 SDK. After the project is created, add Maven support by doing a right-click on the project and choosing **Add Framework Support**. Then select the Maven checkbox and click **OK**.
```

Add a new Kotlin File/Class to the `src/main/kotlin` folder. Set the file name to `app`. The IDE will automatically add the extension `.kt`, which indicates it is a Kotlin file. Add the following code:

```kotlin
package com.okta.developers

fun main() {
    println("Hello, World!")
}
```
In the lines above, you can see a package declaration, which is optional. If not present, everything goes to the default package. The `main` function is the entry point to the application and can be declared without parameters (if your program does not need to accept command line arguments) and return nothing. The `println` function is part of the [Kotlin Standard Library](https://kotlinlang.org/api/latest/jvm/stdlib/index.html) and prints the message and the line separator to the standard output stream.

**NOTE**: In Kotlin, it is not required to match directories and packages. Source files can be placed arbitrarily in the file system, and one file can contain multiple classes. For pure Kotlin projects, the common root package `com.okta.developers` is [omitted in the filesystem](https://kotlinlang.org/docs/reference/coding-conventions.html#source-code-organization).

> Edit your `pom.xml` to add Kotlin dependencies and the **Kotlin Maven Plugin**.

Reviewer: ```suggestion
Edit your `pom.xml` to add Kotlin dependencies and the Kotlin Maven Plugin.
```
This is not something you click on, so I don't think it needs to be bold.

                <executions>
                    <execution>
                        <goals>
                            <goal>java</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <mainClass>${main.class}</mainClass>
                </configuration>
            </plugin>
        </plugins>
    </build>
</project>
```
Build and run application with the following Maven command: 

```txt
mvn package exec:java
```

It should output the hello message:

```txt
Hello, World!
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

### Kotlin Idioms

Now, let's apply some Kotlin [idioms](https://kotlinlang.org/docs/reference/idioms.html) to solve a common interview problem: finding anagrams in substrings. Given a string, find the number of pairs of substrings that are anagrams of each other.

First, update the `main` function in `app.kt` to make it scan the string to analyze from the standard input:

```kotlin
package com.okta.developers

fun main() {

> 

Reviewer: Remove blank line

4
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

## Tutorial: Use Kotlin With Spring Boot Securely

Finally, let's see how easy is to create a Kotlin Spring Boot application with Okta OpenID Connect (OIDC) authentication.

Using the [Spring Initializr API](https://start.spring.io/), create a Maven project with the following command:

```shell
curl https://start.spring.io/starter.zip -d dependencies=web,okta \
-d language=kotlin \
-d type=maven-project \
-d groupId=com.okta.developer \
-d artifactId=kotlin-spring-boot  \
-d name="Kotlin Spring Boot" \
-d description="Demo project of a Kotlin Spring Boot application" \
-d packageName=com.okta.developer \
-o kotlin-spring-boot.zip
```
Unzip the file:

```txt
unzip kotlin-spring-boot.zip -d kotlin-spring-boot
cd kotlin-spring-boot
```

## Secure your Application with OpenID Connect
If you already have an Okta account, see the Create a Web Application in Okta sidebar below. Otherwise, we created a Maven plugin that configures a free Okta developer account + an OIDC app (in under a minute!).

```txt

> ./mvnw com.okta:okta-maven-plugin:setup

Reviewer: NICE!

```

You should see the following output:

```
First name: Jimena
Last name: Garbarino
Email address: ***
Company: ***
Creating new Okta Organization, this may take a minute:
OrgUrl: https://dev-123456.okta.com
Check your email address to verify your account.

Writing Okta SDK config to: /home/indiepopart/.okta/okta.yaml

Configuring a new OIDC, almost done:
Created OIDC application, client-id: ***
```

Check your email and follow the instructions to activate your Okta account.

> **If you already have an Okta Developer account**
> 
> Log in and create a new Application:

> > - From the Applications page, choose Add Application.

Reviewer: ```suggestion
> - From the **Applications** page, choose **Add Application**.
```

> > - On the Create New Application page, select Web.

Reviewer: ```suggestion
> - On the Create New Application page, select **Web**.
```

> > - Give your app a memorable name, add http://localhost:8080/login/oauth2/code/okta and http://localhost:8080/authorization-code/callback as a Login redirect URIs.

Reviewer: ```suggestion
> - Give your app a memorable name, add `http://localhost:8080/login/oauth2/code/okta` as a Login redirect URI.
```

> 
> Copy the issuer (found under **API** > **Authorization Servers**), client ID, and client secret into `src/main/resources/application.properties`.
> 
```properties
okta.oauth2.issuer=$issuer
okta.oauth2.client-id=$clientId
okta.oauth2.client-secret=$clientSecret
```

> Edit `KotlinSpringBootApplication` located the package `com.okta.developer` to add a controller mapping that will print a welcome message in the browser window.

Reviewer: This sentence doesn't make sense.

Reviewer: Should it be "located in the package"?

Author: agreed

```kotlin
package com.okta.developer

import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController

@SpringBootApplication
@RestController
class KotlinSpringBootApplication {

    @GetMapping("/")
    fun hello(@AuthenticationPrincipal authenticationToken: OAuth2AuthenticationToken): String {

> 

Reviewer: Remove blank line.

Reviewer: Another way to do this is to grab the `OidcUser` and call `getFullName()`. Not sure if that's worth mentioning.

        return "Welcome ${authenticationToken.principal.attributes["name"]}"
    }
}

fun main(args: Array<String>) {
    runApplication<KotlinSpringBootApplication>(*args)
}
```

> As you can see, in the `hello` function, the authenticated user is passed in the `authenticationToken`. `@AuthenticationPrincipal` annotation from Spring Security helps to resolve the principal to the expected principal type. The welcome message is built using **String interpolation**, a Kotlin idiom for variable substitution inside strings.

Reviewer: ```suggestion
As you can see, in the `hello()` function, the authenticated user is passed in the `authenticationToken`. The `@AuthenticationPrincipal` annotation from Spring Security helps to resolve the principal to the expected principal type. The welcome message is built using **String interpolation**, a Kotlin idiom for variable substitution inside strings.
```

Run the application with the following command:

> ```txt

Reviewer: ```suggestion
```shell
```

Author: I've had mixed luck with using `shell`, `sh`, and `bash` most commands end up getting colored wrong, i've given up and started using `txt` _unless_ it's an actual shell script

./mvnw spring-boot:run
```

> Browse to`http://localhost:8080` and the application should start an OAuth 2.0  authentication code flow, redirecting to the Okta login page.

Reviewer: ```suggestion
Browse to`http://localhost:8080` and the application should start an OAuth 2.0 authentication code flow, redirecting to the Okta login page.
```

{% img blog/tutorial-kotlin-beginners-guide/okta-sign-in-widget.png alt:"Okta Sign-In Widget screenshot" width:"400" %}{: .center-image }

After the login, you should see the welcome message:

```txt
Welcome, Jimena Garbarino
```

You now have a secure application with just a few lines of Kotlin and the help of Spring Boot!

## Learn More about Kotlin, Java, and Secure Authentication with these Tutorials

> I hope you could grasp how succinct and expressive Kotlin is and why developers are loving it. You can find all the tutorial code in repositories [kotlin-hello-world](https://github.com/oktadeveloper/kotlin-hello-world-example) and [kotlin-spring-boot](https://github.com/oktadeveloper/kotlin-spring-boot-tutorial) on Github. To learn more about Kotlin, check out the following links:

Reviewer: ```suggestion
I hope this blog post helped you grasp how succinct and expressive Kotlin is and why developers are loving it. You can find all the tutorial code in repositories [kotlin-hello-world](https://github.com/oktadeveloper/kotlin-hello-world-example) and [kotlin-spring-boot](https://github.com/oktadeveloper/kotlin-spring-boot-tutorial) on GitHub. To learn more about Kotlin, check out the following links:
```

- [Build a Basic CRUD App in Android with Kotlin
](https://developer.okta.com/blog/2018/09/11/android-kotlin-crud)
- [Build an Application with Spring Boot and Kotlin](https://developer.okta.com/blog/2019/09/17/build-a-spring-boot-kotlin-app)

> - [Kotlin Reference](https://kotlinlang.org/docs/reference/)

Reviewer: Blog post links should be relative and the first one should be on one line, not two.

> Questions? Requests for a future post? Drop them in the comments! And don't forget to follow [@oktadev on Twitter](https://twitter.com/oktadev) and [subscribe on Youtube](http://youtube.com/c/oktadev).

Reviewer: ```suggestion
Questions? Requests for a future post? Drop them in the comments! And don't forget to follow [@oktadev on Twitter](https://twitter.com/oktadev) and [subscribe to our YouTube](http://youtube.com/c/oktadev).
```
